src/slice-assign.c

```diff
@@ -35,7 +35,7 @@ SEXP vec_assign(SEXP x, SEXP index, SEXP value) {
   value = PROTECT(vec_coercible_cast(value, x, &value_arg, &x_arg));
   SEXP value_proxy = PROTECT(vec_proxy(value));
 
-  index = PROTECT(vec_as_index(index, x));
+  index = PROTECT(vec_as_index(index, vec_size(x), vec_names(x)));
```

Comment:
`vec_names()` needs a `PROTECT()` because `Rf_getAttrib()` might return a fresh SEXP. Alternatively we could import the attribute getter implemented in rlang which doesn't allocate (but then you can't get  names vector for pairlists for instance).

Reply:
I've just `PROTECT()`ed in two places for now.